src/core/simd.d

```diff
@@ -621,6 +630,130 @@ version (D_SIMD)
             return cast(V)__simd(XMM.LODDQU, *cast(const void16*)p);
     }
 
+    /*************************************
+    * Store vector to unaligned address.
+    * This is a compiler intrinsic.
+    * Params:
+    *    p = pointer to vector
+    *    value = value to store
+    * Returns:
+    *    value
+    */
+
+    V storeUnaligned(V)(V* p, V value)
+        if (is(V == void16) ||
+            is(V == byte16) ||
+            is(V == ubyte16) ||
+            is(V == short8) ||
+            is(V == ushort8) ||
+            is(V == int4) ||
+            is(V == uint4) ||
+            is(V == long2) ||
+            is(V == ulong2) ||
+            is(V == double2) ||
+            is(V == float4))
+    {
+        pragma(inline, true);
+        static if (is(V == double2))
+            return cast(V)__simd_sto(XMM.STOUPD, *cast(void16*)p, value);
+        else static if (is(V == float4))
+            return cast(V)__simd_sto(XMM.STOUPS, *cast(void16*)p, value);
+        else
+            return cast(V)__simd_sto(XMM.STODQU, *cast(void16*)p, value);
+    }
+}
+else version (LDC)
+{
+    /**
+    Load unaligned vector from address.
+    Params:
+        p = pointer to vector
+    Returns:
+        Vector
+    */
+    pragma(inline, true);
```

Comment:
This statement-form probably doesn't have any effect (should only be used inside a function and then affects that function, but requires semantic analysis of that function), so please get rid of the trailing semicolon.

Reply:
Yes my bad. But just for info, what did you mean by this?
> should only be used inside a function and then affects that function, but requires semantic analysis of that function

What `pragma(inline)` has to do with the semantic analysis of a function?

Reply:
Meaning that if the body of the function isn't analyzed, it won't be inlined.

a.d:
```
void foo()
{
    pragma(inline, true);
}

pragma(inline, true)
void bar() {}
```

main.d:
```
import a;
void main()
{
    foo();
    bar();
}
```

`ldc2 -output-ll main.d`, then viewing `main.ll` with text editor: `foo()` call isn't inlined/eliminated. [It is if `foo` is a template.]

Reply:
Oh now I see. I guess LDC analyzes functions that have at least one statement. I don't know how a `pragma` is called formally, but I guess it's not an actual statement.
And so, `foo` has nothing and so it's body is not analyzed (hence the compiler never sees the `pragma`). `bar` is not analyzed either, but the `pragma` is seen because it's not part of its body.

Is this LDC specific? Also, the reason that you put them in separate files is because in one file, it's one "compilation unit" and so the compiler can reason and just not declare any of them? While I don't know, on a separate it has to declare `foo` as external symbol.

Reply:
If it's all in the same file, then the foo-body is fully analyzed (=> codegen into object file), and then the pragma is applied, and all is well. If you're only compiling `main.d`, then `foo` isn't codegen'd, the frontend doesn't analyze more than it has to and misses the pragma in the body. And if you compile both modules into one object file (`ldc2 -output-ll -singleobj a.d main.d`), all is well again. And if you omit the `-singleobj` and compile the 2 modules into 2 separate object files in a single cmdline, I've just seen that both calls aren't inlined... => bug ;)

Reply:
Give me a second as I'm trying to comprehend this. To rephrase:

- If I try to compile a single file, then a single object file is generated. And the front-end has to analyze all the symbols, as they're in the same file. So everything ok.
- Now, if compile `main.d` _only_, the compiler has to generate one object file for `main.d`. That object file calls some functions `foo` and `bar` which though will be in some _other_ object file and the only thing that the front-end cares about is the API of these, not the body. So, it misses the body.
- The bug is that in 2 object files, only `bar` should have been inlined? As if just compiling `main.d`?